src/polynomial.cpp

```diff
@@ -0,0 +1,297 @@
+#include "mul.h"
+#include "pow.h"
+#include "constants.h"
+
+namespace SymEngine {
+
+UnivariatePolynomial::UnivariatePolynomial(const RCP<const Symbol> &var, const uint &degree, map_uint_mpz&& dict) :
+     degree_{degree}, var_{var}, dict_{std::move(dict)} {
+
+    SYMENGINE_ASSERT(is_canonical(degree_, dict_))
+}
+
+UnivariatePolynomial::UnivariatePolynomial(const RCP<const Symbol> &var, const std::vector<mpz_class> &v) :
+     var_{var} {
+
+    for (uint i = 0; i < v.size(); i++) {
+        if (v[i] != 0)
+            dict_add_term(dict_, v[i], i);
+    }
+    degree_ = v.size() - 1;
+
+    SYMENGINE_ASSERT(is_canonical(degree_, dict_))
+}
+
+bool UnivariatePolynomial::is_canonical(const uint &degree_, const map_uint_mpz& dict)
+{
+    map_uint_mpz ordered(dict.begin(), dict.end());
+    uint prev_degree = (--ordered.end())->first;
+    if (prev_degree != degree_)
+        return false;
+
+    return true;
+}
+
+std::size_t UnivariatePolynomial::__hash__() const 
+{
+    std::hash<std::string> hash_string;
+    std::size_t seed = UNIVARIATEPOLYNOMIAL;
+
+    seed += hash_string(this->var_->get_name());
+    for (auto &it : this->dict_)
+    {   
+        std::size_t temp = UNIVARIATEPOLYNOMIAL;
+        hash_combine<uint>(temp, it.first);
+        hash_combine<long long int>(temp, it.second.get_si());
+        seed += temp;
+    }
+    return seed;
+}
+
+bool UnivariatePolynomial::__eq__(const Basic &o) const
+{
+    if (eq(var_, static_cast<const UnivariatePolynomial &>(o).var_) &&
+        map_uint_mpz_eq(dict_, static_cast<const UnivariatePolynomial &>(o).dict_))
+        return true;
+
+    return false;
+}
+
+int UnivariatePolynomial::compare(const Basic &o) const
+{
+    const UnivariatePolynomial &s = static_cast<const UnivariatePolynomial &>(o);
+
+    if (dict_.size() != s.dict_.size())
+        return (dict_.size() < s.dict_.size()) ? -1 : 1;
+
+    int cmp = var_->compare(*s.var_);
+    if (cmp != 0)
+        return cmp;
+
+    return map_uint_mpz_compare(dict_, s.dict_);
+}
+
+RCP<const Basic> UnivariatePolynomial::from_dict(const RCP<const Symbol> &var, map_uint_mpz &&d)
+{
+    if (d.size() == 1) {
+        if (d.begin()->first == 0) 
+            return integer(d.begin()->second);
+        else if (d.begin()->first == 1) {
+            if (d.begin()->second == 0)
+                return zero;
+            else if (d.begin()->second == 1) 
+                return var;
+            else 
+                return Mul::from_dict(integer(d.begin()->second), {{var, one}});
+        } else {
+            if (d.begin()->second == 0)
+                return zero;
+            else if (d.begin()->second == 1) 
+                return pow(var, integer(d.begin()->first));
+            else 
+                return Mul::from_dict(integer(d.begin()->second), 
+                    {{var, integer(d.begin()->first)}});
+        }
+    } else {
+        return rcp(new UnivariatePolynomial(var, (--(d.end()))->first, std::move(d)));
```

Comment:
Yes, `from_dict` is used to construct `Polynomial` when we do not know the degree.
It uses the fact the std::map is ordered.